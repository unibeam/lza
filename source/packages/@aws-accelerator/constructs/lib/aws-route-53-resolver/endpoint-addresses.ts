/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as path from 'path';

export interface IEndpointAddresses extends cdk.IResource {
  /**
   * The IP addresses of the endpoint.
   */
  readonly ipAddresses: cdk.Reference;
}

export interface EndpointAddressesProps {
  /**
   * The ID of the Route 53 Resolver endpoint.
   */
  readonly endpointId: string;
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly kmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;
}

export class EndpointAddresses extends cdk.Resource implements IEndpointAddresses {
  public ipAddresses: cdk.Reference;

  constructor(scope: Construct, id: string, props: EndpointAddressesProps) {
    super(scope, id);

    const provider = cdk.CustomResourceProvider.getOrCreateProvider(this, 'Custom::ResolverEndpointAddresses', {
      codeDirectory: path.join(__dirname, 'get-endpoint-addresses/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
      policyStatements: [
        {
          Effect: 'Allow',
          Action: ['route53resolver:ListResolverEndpointIpAddresses'],
          Resource: '*',
        },
      ],
    });

    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: 'Custom::ResolverEndpointAddresses',
      serviceToken: provider.serviceToken,
      properties: {
        endpointId: props.endpointId,
        region: cdk.Stack.of(this).region,
      },
    });

    /**
     * Singleton pattern to define the log group for the singleton function
     * in the stack
     */
    const stack = cdk.Stack.of(scope);
    const logGroup =
      (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
      new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
        logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
        retention: props.logRetentionInDays,
        encryptionKey: props.kmsKey,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });
    resource.node.addDependency(logGroup);

    this.ipAddresses = resource.getAtt('ipAddresses');
  }
}
