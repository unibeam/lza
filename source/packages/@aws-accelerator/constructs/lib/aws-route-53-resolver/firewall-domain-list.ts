/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as path from 'path';

export interface IResolverFirewallDomainList extends cdk.IResource {
  /**
   * The ID of the domain list.
   */
  readonly listId: string;

  /**
   * The name of the domain list.
   */
  readonly name: string;

  /**
   * The Amazon Resource Name (ARN) of the firewall domain list.
   */
  readonly listArn?: string;
}

export enum ResolverFirewallDomainListType {
  CUSTOM = 'CUSTOM',
  MANAGED = 'MANAGED',
}

export interface ResolverFirewallDomainListProps {
  /**
   * The name of the domain list.
   */
  readonly name: string;

  /**
   * The type of the domain list.
   */
  readonly type: ResolverFirewallDomainListType;

  /**
   * Path to a file containing a domain list.
   */
  readonly path?: string;

  /**
   * A list of CloudFormation tags
   */
  readonly tags?: cdk.CfnTag[];

  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly kmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;
}

export class ResolverFirewallDomainList extends cdk.Resource implements IResolverFirewallDomainList {
  public readonly listId: string;
  public readonly name: string;
  public readonly listArn?: string;
  private assetUrl?: string;

  constructor(scope: Construct, id: string, props: ResolverFirewallDomainListProps) {
    super(scope, id);

    this.name = props.name;

    if (props.type === ResolverFirewallDomainListType.CUSTOM) {
      // Check if path was provided
      if (!props.path) {
        throw new Error('path property must be specified when creating domain list of type CUSTOM');
      }
      this.assetUrl = this.getAssetUrl(props.path);

      props.tags?.push({ key: 'Name', value: this.name });

      // Create custom domain list with uploaded asset file
      const resource = new cdk.aws_route53resolver.CfnFirewallDomainList(this, 'Resource', {
        domainFileUrl: this.assetUrl,
        tags: props.tags,
      });

      this.listArn = resource.attrArn;
      this.listId = resource.attrId;
    } else {
      // Create custom resource provider
      const RESOURCE_TYPE = 'Custom::ResolverManagedDomainList';

      const provider = cdk.CustomResourceProvider.getOrCreateProvider(this, RESOURCE_TYPE, {
        codeDirectory: path.join(__dirname, 'get-domain-lists/dist'),
        runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
        policyStatements: [
          {
            Effect: 'Allow',
            Action: ['route53resolver:ListFirewallDomainLists'],
            Resource: '*',
          },
        ],
      });

      // Get managed domain list ID
      const resource = new cdk.CustomResource(this, 'Resource', {
        resourceType: RESOURCE_TYPE,
        serviceToken: provider.serviceToken,
        properties: {
          listName: props.name,
          region: cdk.Stack.of(this).region,
        },
      });

      /**
       * Singleton pattern to define the log group for the singleton function
       * in the stack
       */
      const stack = cdk.Stack.of(scope);
      const logGroup =
        (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
        new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
          logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
          retention: props.logRetentionInDays,
          encryptionKey: props.kmsKey,
          removalPolicy: cdk.RemovalPolicy.DESTROY,
        });
      resource.node.addDependency(logGroup);

      this.listId = resource.ref;
    }
  }

  private getAssetUrl(assetUrlPath: string): string {
    const asset = new cdk.aws_s3_assets.Asset(this, 'Asset', {
      path: assetUrlPath,
    });
    return asset.s3ObjectUrl;
  }
}
