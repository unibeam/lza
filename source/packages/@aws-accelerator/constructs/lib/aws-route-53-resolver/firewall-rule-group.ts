/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface IResolverFirewallRuleGroup extends cdk.IResource {
  /**
   * The ARN (Amazon Resource Name) of the rule group.
   */
  readonly groupArn: string;

  /**
   * The ID of the rule group.
   */
  readonly groupId: string;

  /**
   * The name of the rule group.
   */
  readonly name: string;
}

export interface ResolverFirewallRuleGroupProps {
  /**
   * A list of the rules that you have defined.
   */
  readonly firewallRules: cdk.aws_route53resolver.CfnFirewallRuleGroup.FirewallRuleProperty[];

  /**
   * The name of the rule group.
   */
  readonly name: string;

  /**
   * A list of CloudFormation tags.
   */
  readonly tags?: cdk.CfnTag[];
}

export class ResolverFirewallRuleGroup extends cdk.Resource implements IResolverFirewallRuleGroup {
  public readonly groupArn: string;
  public readonly groupId: string;
  public readonly name: string;

  constructor(scope: Construct, id: string, props: ResolverFirewallRuleGroupProps) {
    super(scope, id);

    this.name = props.name;
    props.tags?.push({ key: 'Name', value: this.name });

    const resource = new cdk.aws_route53resolver.CfnFirewallRuleGroup(this, 'Resource', {
      firewallRules: props.firewallRules,
      tags: props.tags,
    });

    this.groupArn = resource.attrArn;
    this.groupId = resource.ref;
  }
}

export interface ResolverFirewallRuleGroupAssociationProps {
  /**
   * The unique identifier of the firewall rule group.
   */
  readonly firewallRuleGroupId: string;

  /**
   * The setting that determines the processing order of the rule group
   * among the rule groups that are associated with a single VPC.
   */
  readonly priority: number;

  /**
   * The unique identifier of the VPC that is associated with the rule group.
   */
  readonly vpcId: string;

  /**
   * If enabled, this setting disallows modification or removal of the association
   */
  readonly mutationProtection?: string;

  /**
   * A list of CloudFormation tags.
   */
  readonly tags?: cdk.CfnTag[];
}

export class ResolverFirewallRuleGroupAssociation extends cdk.Resource {
  constructor(scope: Construct, id: string, props: ResolverFirewallRuleGroupAssociationProps) {
    super(scope, id);

    new cdk.aws_route53resolver.CfnFirewallRuleGroupAssociation(this, 'Resource', {
      firewallRuleGroupId: props.firewallRuleGroupId,
      priority: props.priority,
      vpcId: props.vpcId,
      mutationProtection: props.mutationProtection,
      tags: props.tags,
    });
  }
}
