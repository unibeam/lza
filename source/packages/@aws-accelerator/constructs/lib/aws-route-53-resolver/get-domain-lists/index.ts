/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as AWS from 'aws-sdk';

import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';
AWS.config.logger = console;

/**
 * Get Route 53 resolver endpoint details - Lambda handler
 *
 * @param event
 * @returns
 */
export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      PhysicalResourceId: string;
      Status: string;
    }
  | undefined
> {
  const region: string = event.ResourceProperties['region'];
  const listName: string = event.ResourceProperties['listName'];
  const solutionId = process.env['SOLUTION_ID'];
  const resolverClient = new AWS.Route53Resolver({ region: region, customUserAgent: solutionId });

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      let nextToken: string | undefined = undefined;
      let resourceId: string | undefined = undefined;
      do {
        const page = await throttlingBackOff(() =>
          resolverClient.listFirewallDomainLists({ NextToken: nextToken }).promise(),
        );

        // Loop through IP addresses and push to array
        for (const item of page.FirewallDomainLists ?? []) {
          if (item.Name === listName) {
            resourceId = item.Id;
          }
        }
        nextToken = page.NextToken;
      } while (nextToken);

      if (!resourceId) {
        throw new Error(`Managed domain list ${listName} does not exist.`);
      }

      return {
        PhysicalResourceId: resourceId,
        Status: 'SUCCESS',
      };

    case 'Delete':
      // Do nothing
      return {
        PhysicalResourceId: event.PhysicalResourceId,
        Status: 'SUCCESS',
      };
  }
}
