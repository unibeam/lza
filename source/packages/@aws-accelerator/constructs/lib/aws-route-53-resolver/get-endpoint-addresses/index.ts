/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as AWS from 'aws-sdk';

import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';
AWS.config.logger = console;

/**
 * Get Route 53 resolver endpoint details - Lambda handler
 *
 * @param event
 * @returns
 */
export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      PhysicalResourceId: string;
      Data: {
        ipAddresses: { Ip: string }[];
      };
      Status: string;
    }
  | {
      PhysicalResourceId: string;
      Status: string;
    }
  | undefined
> {
  const region: string = event.ResourceProperties['region'];
  const endpointId: string = event.ResourceProperties['endpointId'];
  const solutionId = process.env['SOLUTION_ID'];
  const resolverClient = new AWS.Route53Resolver({ region: region, customUserAgent: solutionId });

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      const ipArray: { Ip: string }[] = [];
      let nextToken: string | undefined = undefined;
      do {
        const page = await throttlingBackOff(() =>
          resolverClient
            .listResolverEndpointIpAddresses({ ResolverEndpointId: endpointId, NextToken: nextToken })
            .promise(),
        );

        // Loop through IP addresses and push to array
        for (const item of page.IpAddresses ?? []) {
          if (item.Ip) {
            ipArray.push({ Ip: item.Ip });
          }
        }

        nextToken = page.NextToken;
      } while (nextToken);

      return {
        PhysicalResourceId: endpointId,
        Data: {
          ipAddresses: ipArray,
        },
        Status: 'SUCCESS',
      };

    case 'Delete':
      // Do nothing
      return {
        PhysicalResourceId: event.PhysicalResourceId,
        Status: 'SUCCESS',
      };
  }
}
