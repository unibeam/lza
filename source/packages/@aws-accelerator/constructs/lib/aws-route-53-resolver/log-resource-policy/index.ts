/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as AWS from 'aws-sdk';
import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';
AWS.config.logger = console;

/**
 * log-resource-policy - Lambda handler
 *
 * @param event
 * @returns
 */
export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      PhysicalResourceId: string;
      Status: string;
    }
  | undefined
> {
  interface ResourceLogPolicy {
    policyName: string;
    policyStatements: [];
  }

  const resourceLogPolicy = event.ResourceProperties as unknown as ResourceLogPolicy;
  const policyName = resourceLogPolicy.policyName;
  const policyDocument = {
    Version: '2012-10-17',
    Statement: resourceLogPolicy.policyStatements,
  };
  console.log(policyDocument);

  const logsClient = new AWS.CloudWatchLogs();

  switch (event.RequestType) {
    case 'Update':
    case 'Create':
      console.log(`Creating CloudWatch log resource policy.`);
      await throttlingBackOff(() =>
        logsClient
          .putResourcePolicy({
            policyName: policyName,
            policyDocument: JSON.stringify(policyDocument),
          })
          .promise(),
      );
      return {
        PhysicalResourceId: policyName,
        Status: 'SUCCESS',
      };
    case 'Delete':
      console.log(`Deleting CloudWatch log resource policy.`);
      await throttlingBackOff(() => logsClient.deleteResourcePolicy({ policyName: policyName }).promise());
      return {
        PhysicalResourceId: policyName,
        Status: 'SUCCESS',
      };
  }
}
