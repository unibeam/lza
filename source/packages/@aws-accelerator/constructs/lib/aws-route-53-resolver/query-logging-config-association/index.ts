/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as AWS from 'aws-sdk';

import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';
AWS.config.logger = console;

/**
 * query-logging-config-association - Lambda handler
 *
 * @param event
 * @returns
 */
export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      PhysicalResourceId: string | undefined;
      Status: string;
    }
  | undefined
> {
  interface ResolverQueryLogConfigAssociation {
    ResolverQueryLogConfigId: string;
    VpcId: string;
  }

  const resolverQueryLogConfigAssociation = event.ResourceProperties as unknown as ResolverQueryLogConfigAssociation;

  const { ResolverQueryLogConfigId, VpcId } = resolverQueryLogConfigAssociation;
  const route53ResolverClient = new AWS.Route53Resolver();

  switch (event.RequestType) {
    case 'Update':
    case 'Create':
      console.log(`Associating Route53 resolver query log config ${ResolverQueryLogConfigId} to VPC ${VpcId}`);
      const data = await throttlingBackOff(() =>
        route53ResolverClient
          .associateResolverQueryLogConfig({
            ResolverQueryLogConfigId: ResolverQueryLogConfigId,
            ResourceId: VpcId,
          })
          .promise(),
      );
      return {
        PhysicalResourceId: data.ResolverQueryLogConfigAssociation?.Id,
        Status: 'SUCCESS',
      };
    case 'Delete':
      console.log(`Disassociating Route53 resolver query log config ${ResolverQueryLogConfigId} to VPC ${VpcId}`);
      await throttlingBackOff(() =>
        route53ResolverClient
          .disassociateResolverQueryLogConfig({
            ResolverQueryLogConfigId: ResolverQueryLogConfigId,
            ResourceId: VpcId,
          })
          .promise(),
      );
      return {
        PhysicalResourceId: event.PhysicalResourceId,
        Status: 'SUCCESS',
      };
  }
}
