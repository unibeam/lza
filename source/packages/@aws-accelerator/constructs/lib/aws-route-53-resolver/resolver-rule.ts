/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

import { EndpointAddresses } from './endpoint-addresses';

export interface IResolverRule extends cdk.IResource {
  /**
   * The friendly name of the resolver rule.
   */
  readonly name: string;

  /**
   * The Amazon Resource Name (ARN) of the resolver rule.
   */
  readonly ruleArn: string;

  /**
   * The ID that Resolver assigned to the resolver rule when you created it.
   */
  readonly ruleId: string;
}

export interface ResolverRuleProps {
  /**
   * DNS queries for this domain name are forwarded to the IP addresses that are specified in `TargetIps`.
   */
  readonly domainName: string;

  /**
   * The name for the Resolver rule.
   */
  readonly name: string;

  /**
   * The ID of the endpoint that the rule is associated with.
   */
  readonly resolverEndpointId?: string;

  /**
   * The type of resolver rule: FORWARD, RECURSIVE, or SYSTEM.
   *
   * @default FORWARD
   */
  readonly ruleType?: string;

  /**
   * Choose to target an inbound resolver endpoint for name resolution.
   */
  readonly targetInbound?: string;

  /**
   * An array that contains the IP addresses and ports that an outbound endpoint forwards DNS queries to.
   */
  readonly targetIps?: cdk.aws_route53resolver.CfnResolverRule.TargetAddressProperty[];

  /**
   * A list of CloudFormation tags.
   */
  readonly tags?: cdk.CfnTag[];

  /**
   * Custom resource lambda log group encryption key
   */
  readonly kmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays?: number;
}

export class ResolverRule extends cdk.Resource implements IResolverRule {
  public readonly name: string;
  public readonly ruleArn: string;
  public readonly ruleId: string;
  private targetIps?: cdk.aws_route53resolver.CfnResolverRule.TargetAddressProperty[] | cdk.Reference;

  constructor(scope: Construct, id: string, props: ResolverRuleProps) {
    super(scope, id);

    this.name = props.name;

    if (props.targetInbound) {
      if (!props.logRetentionInDays) {
        throw new Error(`logRetentionInDays property must be included if targetInbound property is defined.`);
      }
      this.targetIps = this.lookupInbound(props.targetInbound, props.logRetentionInDays, props.kmsKey);
    } else {
      this.targetIps = props.targetIps;
    }

    const resource = new cdk.aws_route53resolver.CfnResolverRule(this, 'Resource', {
      domainName: props.domainName,
      resolverEndpointId: props.ruleType === 'SYSTEM' ? undefined : props.resolverEndpointId,
      ruleType: props.ruleType ?? 'FORWARD',
      name: this.name,
      targetIps: this.targetIps,
      tags: props.tags,
    });
    cdk.Tags.of(this).add('Name', this.name);

    this.ruleArn = resource.attrArn;
    this.ruleId = resource.attrResolverRuleId;
  }

  private lookupInbound(endpointId: string, logRetentionInDays: number, kmsKey?: cdk.aws_kms.IKey): cdk.Reference {
    const lookup = new EndpointAddresses(this, 'LookupInbound', {
      endpointId: endpointId,
      kmsKey,
      logRetentionInDays,
    });
    return lookup.ipAddresses;
  }
}

export interface ResolverRuleAssociationProps {
  /**
   * The ID of the Resolver rule to associate.
   */
  readonly resolverRuleId: string;

  /**
   * The ID of the VPC to associate.
   */
  readonly vpcId: string;
}

export class ResolverRuleAssociation extends cdk.Resource {
  constructor(scope: Construct, id: string, props: ResolverRuleAssociationProps) {
    super(scope, id);

    new cdk.aws_route53resolver.CfnResolverRuleAssociation(this, 'Resource', {
      resolverRuleId: props.resolverRuleId,
      vpcId: props.vpcId,
    });
  }
}
