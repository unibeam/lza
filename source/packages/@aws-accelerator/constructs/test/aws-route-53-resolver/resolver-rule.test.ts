/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { ResolverRule, ResolverRuleAssociation } from '../../lib/aws-route-53-resolver/resolver-rule';
import { snapShotTest } from '../snapshot-test';
import { describe, expect, it } from '@jest/globals';

const testNamePrefix = 'Construct(ResolverRule): ';

const forwardRuleStack = new cdk.Stack();
const systemRuleStack = new cdk.Stack();

const ipAddresses = [{ ip: '1.1.1.1' }, { ip: '2.2.2.2' }];

const forwardRule = new ResolverRule(forwardRuleStack, 'TestResolverRule', {
  domainName: 'test.com',
  name: 'TestResolverRule',
  resolverEndpointId: 'TestEndpoint',
  targetIps: ipAddresses,
  tags: [],
  kmsKey: new cdk.aws_kms.Key(forwardRuleStack, 'CustomKey', {}),
  logRetentionInDays: 3653,
});

new ResolverRuleAssociation(forwardRuleStack, 'TestResolverRuleAssoc', {
  resolverRuleId: forwardRule.ruleId,
  vpcId: 'TestVpc',
});

const systemRule = new ResolverRule(systemRuleStack, 'TestResolverRule', {
  domainName: 'test.com',
  name: 'TestResolverRule',
  resolverEndpointId: 'TestEndpoint',
  targetIps: ipAddresses,
  tags: [],
  kmsKey: new cdk.aws_kms.Key(systemRuleStack, 'CustomKey', {}),
  logRetentionInDays: 3653,
  ruleType: 'SYSTEM',
});

new ResolverRuleAssociation(systemRuleStack, 'TestResolverRuleAssoc', {
  resolverRuleId: systemRule.ruleId,
  vpcId: 'TestVpc',
});

describe('ResolverRule', () => {
  snapShotTest(testNamePrefix, forwardRuleStack);
  snapShotTest(testNamePrefix, systemRuleStack);
  it('throw error when targetInbound is specified without logRetention', () => {
    function targetInboundLogRetentionError() {
      new ResolverRule(systemRuleStack, 'TargetInboundLogRetentionError', {
        domainName: 'test.com',
        name: 'TestResolverRule',
        resolverEndpointId: 'TestEndpoint',
        targetIps: ipAddresses,
        tags: [],
        targetInbound: 'targetInbound',
        kmsKey: new cdk.aws_kms.Key(systemRuleStack, 'CustomKeyTargetInboundLogRetentionError', {}),
        ruleType: 'SYSTEM',
      });
    }
    expect(targetInboundLogRetentionError).toThrow(
      new Error('logRetentionInDays property must be included if targetInbound property is defined.'),
    );
  });
  it('test private function lookup inbound', () => {
    const testLookupInbound = new ResolverRule(systemRuleStack, 'TestLookupInbound', {
      domainName: 'test.com',
      name: 'TestResolverRule',
      resolverEndpointId: 'TestEndpoint',
      targetIps: ipAddresses,
      tags: [],
      targetInbound: 'targetInbound',
      kmsKey: new cdk.aws_kms.Key(systemRuleStack, 'CustomKeyTestLookupInbound', {}),
      logRetentionInDays: 3653,
      ruleType: 'SYSTEM',
    });
    //output of testLookupId ruleId is a cdk token hash so the check here is to make sure its a string
    expect(typeof testLookupInbound.ruleId).toBe('string');
  });
});
