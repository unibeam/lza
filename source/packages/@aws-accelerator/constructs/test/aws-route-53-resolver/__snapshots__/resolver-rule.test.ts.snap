// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResolverRule Construct(ResolverRule):  Snapshot Test 1`] = `
{
  "Resources": {
    "CustomKey1E6D0D07": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "TestResolverRule183FBE0C": {
      "Properties": {
        "DomainName": "test.com",
        "Name": "TestResolverRule",
        "ResolverEndpointId": "TestEndpoint",
        "RuleType": "FORWARD",
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestResolverRule",
          },
        ],
        "TargetIps": [
          {
            "Ip": "1.1.1.1",
          },
          {
            "Ip": "2.2.2.2",
          },
        ],
      },
      "Type": "AWS::Route53Resolver::ResolverRule",
    },
    "TestResolverRuleAssoc7E0DCDC2": {
      "Properties": {
        "ResolverRuleId": {
          "Fn::GetAtt": [
            "TestResolverRule183FBE0C",
            "ResolverRuleId",
          ],
        },
        "VPCId": "TestVpc",
      },
      "Type": "AWS::Route53Resolver::ResolverRuleAssociation",
    },
  },
}
`;

exports[`ResolverRule Construct(ResolverRule):  Snapshot Test 2`] = `
{
  "Resources": {
    "CustomKey1E6D0D07": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "TestResolverRule183FBE0C": {
      "Properties": {
        "DomainName": "test.com",
        "Name": "TestResolverRule",
        "RuleType": "SYSTEM",
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestResolverRule",
          },
        ],
        "TargetIps": [
          {
            "Ip": "1.1.1.1",
          },
          {
            "Ip": "2.2.2.2",
          },
        ],
      },
      "Type": "AWS::Route53Resolver::ResolverRule",
    },
    "TestResolverRuleAssoc7E0DCDC2": {
      "Properties": {
        "ResolverRuleId": {
          "Fn::GetAtt": [
            "TestResolverRule183FBE0C",
            "ResolverRuleId",
          ],
        },
        "VPCId": "TestVpc",
      },
      "Type": "AWS::Route53Resolver::ResolverRuleAssociation",
    },
  },
}
`;
